### PR TITLE
Hardening/valgrind only for debug compilations

### PR DESCRIPTION
**Pull request by kzangeli**
Lots of valgrind problems solved.
First of all the 'exit code 11'. This error occurs when we run valgrind on a 'non-debug-compiled' broker. Non-debug means that "/exit" is not recognized and without that the valgrind test suite cannot function.

Apart from that a few new functions implemented to take care of conditional sleeps for valgrind.
That and a few more tricks has gotten the number of failed functests  (under valgrind that completely changed the timing) to 3:

   subscription_sequence_ontimeinterval.test
   subscription_sequence_wildcards_ontimeinterval.test
   coap_basic.test

coap_basic.test almost works. Of its four part tests three are working now (before, all four failed).
subscription_sequence_ontimeinterval.test and subscription_sequence_wildcards_ontimeinterval.test are pretty much impossible to make work.
We could or in a '.*' in the case of valgrind, that's all I can think of.
Or completely change the test to not depend on exact seconds ...

REGEX((0|1)), REGEX((2|3)), etc, etc.

At least, now almost everything is working.
And thanks to the function "accumulatorDump" not so many failing functions will be added in the future.

### CODE REVIEW

File: test/functionalTest/harnessFunctions.sh [hunk 1]
```diff
@@ -428,6 +443,52 @@ function accumulatorStart()
...
 
+
+
+# ------------------------------------------------------------------------------
+#
+# accumulatorDump
+#
+function accumulatorDump()
+{
+  if [ "$VALGRIND" == "1" ]
```
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: Why don't using valgrindSleep here?

File: test/functionalTest/harnessFunctions.sh [hunk 2]
```diff
@@ -428,6 +443,52 @@ function accumulatorStart()
...
+}
+
+
+# ------------------------------------------------------------------------------
+#
+# accumulator2Dump
+#
+function accumulator2Dump()
+{
+  if [ "$VALGRIND" == "1" ]
```
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: Why don't using valgrindSleep here?

File: test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test [hunk 3]
```diff
@@ -137,13 +137,15 @@ echo
 
 echo "5. Dump accumulator t_01"
 echo "+++++++++++++++++++++++++"
-curl localhost:${LISTENER_PORT}/dump --silent
+#curl localhost:${LISTENER_PORT}/dump --silent
```
Comment (fgalan) [inline test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test]: Leftover comment?

File: test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test [hunk 4]
```diff
@@ -137,13 +137,15 @@ echo
...
+accumulatorDump
 echo
 
 
 echo "6. Dump accumulator t_02"
 echo "+++++++++++++++++++++++++"
-curl localhost:${LISTENER2_PORT}/dump --silent
+#curl localhost:${LISTENER2_PORT}/dump --silent
```
Comment (fgalan) [inline test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test]: Leftover comment?

File: test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test [hunk 5]
```diff
@@ -94,8 +94,8 @@ echo
 
 echo "03. Check that the X-Auth-Token from the subscribe request was propagated to accumulator"
 echo "========================================================================================"
-sleep 1
-curl localhost:${LISTENER_PORT}/dump -s -S
+#sleep 1
```
Comment (fgalan) [inline test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test]: Leftover comment?

File: test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test [hunk 6]
```diff
@@ -131,8 +131,8 @@ echo
 
 echo "06. Check that the X-Auth-Token from the update subscription request was propagated to accumulator"
 echo "=================================================================================================="
-sleep 1
-curl localhost:${LISTENER_PORT}/dump -s -S
+#sleep 1
```
Comment (fgalan) [inline test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test]: Leftover comment?

File: test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test [hunk 7]
```diff
@@ -124,8 +124,8 @@ echo
 
 echo "04. Check that the X-Auth-Token from the subscribe request was propagated to accumulator"
 echo "========================================================================================"
-sleep 1
-curl localhost:${LISTENER_PORT}/dump -s -S
+#sleep 1
```
Comment (fgalan) [inline test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test]: Leftover comment?

File: test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test [hunk 8]
```diff
@@ -161,8 +161,8 @@ echo
 
 echo "07. Make sure the accumulator receives the Update, with the correct X-Auth token"
 echo "================================================================================"
-sleep 1
-curl localhost:${LISTENER_PORT}/dump -s -S
+# sleep 1
```
Comment (fgalan) [inline test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test]: Leftover comment?

File: test/functionalTest/cases/673_xauth_token/subscription.test [hunk 9]
```diff
@@ -85,8 +85,8 @@ echo
 
 echo "03. Check that the X-Auth-Token was propagated to accumulator"
 echo "============================================================="
-sleep 1
-curl localhost:${LISTENER_PORT}/dump -s -S
+#sleep 1
```
Comment (fgalan) [inline test/functionalTest/cases/673_xauth_token/subscription.test]: Leftover comment?

File: test/functionalTest/cases/674_servicepath_in_notifications/service_path_in_ontimeinterval_notification.test [hunk 10]
```diff
@@ -106,8 +106,8 @@ echo
 
 echo "05. Query accumulator"
 echo "====================="
-sleep 2
-curl localhost:${LISTENER_PORT}/dump -s -S
+# sleep 2
```
Comment (fgalan) [inline test/functionalTest/cases/674_servicepath_in_notifications/service_path_in_ontimeinterval_notification.test]: Leftover comment?

File: test/functionalTest/harnessFunctions.sh [hunk 11]
```diff
@@ -428,6 +443,52 @@ function accumulatorStart()
   done
 }
 
+
+
+# ------------------------------------------------------------------------------
+#
+# accumulatorDump
+#
+function accumulatorDump()
```
Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: Nice idea this accumulator*Dump function!

Comment (fgalan) [inline test/functionalTest/harnessFunctions.sh]: NTC

File: test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test [hunk 12]
```diff
@@ -124,6 +124,7 @@ orionCurl --url "$url" --payload "$payload" --host "[::1]" -g 2> /dev/null
 
 echo "6: ++++++++++++++++++++"
 #Get accumulated data
+valgrindSleep 2
 curl -g [::1]:${LISTENER2_PORT}/dump -s -S 2> /dev/null
```
Comment (fgalan) [inline test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test]: It's a pitty cannot use accumulatorDump function for this due to the weird IPv6 syntax... Maybe a parameter could be used?

```
accumulatorDump IPv6
```

Not sure, but I think I remember we have used that technique some time (maybe in accumulatorStart function?)

File: test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test]: I wasn't 100% sure about this.
What does --silent do?

File: test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test]: Same here. Why was this sleep here?
Is it safe to remove it?

File: test/functionalTest/cases/674_servicepath_in_notifications/service_path_in_ontimeinterval_notification.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/674_servicepath_in_notifications/service_path_in_ontimeinterval_notification.test]: It's the exact same thing.
Not sure it is safe to remove these sleeps (that was not depending on valgrind.
A few tests are not enough to assure that it really works

File: test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test (hunk 12 above)
Comment (kzangeli) [inline test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test]: Yes, we could do this, of course. It's brokerStart that has the ipv6 parameter

Comment (kzangeli) [inline test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/674_servicepath_in_notifications/service_path_in_ontimeinterval_notification.test (hunk 10 above)
Comment (kzangeli) [inline test/functionalTest/cases/674_servicepath_in_notifications/service_path_in_ontimeinterval_notification.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/673_xauth_token/subscription.test (hunk 9 above)
Comment (kzangeli) [inline test/functionalTest/cases/673_xauth_token/subscription.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test (hunk 8 above)
Comment (kzangeli) [inline test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test (hunk 7 above)
Comment (kzangeli) [inline test/functionalTest/cases/673_xauth_token/auth_token_three_steps.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/673_xauth_token/auth_token_from_subscribe_request.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test]: Fixed in c7b7a2a

File: test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/322_multitenancy/http_tenant_subscribe_and_update.test]: Fixed in c7b7a2a

File: test/functionalTest/harnessFunctions.sh (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/harnessFunctions.sh]: Fixed in c7b7a2a

File: test/functionalTest/harnessFunctions.sh (hunk 1 above)
Comment (kzangeli) [inline test/functionalTest/harnessFunctions.sh]: Fixed in c7b7a2a

File: test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test [hunk 13]
```diff
@@ -124,11 +124,13 @@ orionCurl --url "$url" --payload "$payload" --host "[::1]" -g 2> /dev/null
 
 echo "6: ++++++++++++++++++++"
 #Get accumulated data
-curl -g [::1]:${LISTENER2_PORT}/dump -s -S 2> /dev/null
+valgrindSleep 2
```
Comment (fgalan) [inline test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test]: accumulator2Dump already includes a `valgrindSleep 2`...

Comment (kzangeli) [inline test/functionalTest/cases/000_ipv6_support/ipv4_ipv6_both.test]: Fixed in 9eaef95

**Comment by fgalan** [comment]
LGTM

**Pull request merged**